hardinfo/dmi_util.c

```diff
@@ -0,0 +1,152 @@
+ *    the Free Software Foundation, version 2.
+ *
+ *    This program is distributed in the hope that it will be useful,
+ *    but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *    GNU General Public License for more details.
+ *
+ *    You should have received a copy of the GNU General Public License
+ *    along with this program; if not, write to the Free Software
+ *    Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA
+ */
+
+#include "hardinfo.h"
+#include "dmi_util.h"
+
+const char *dmi_sysfs_root() {
+    char *candidates[] = {
+        "/sys/devices/virtual/dmi",
+        "/sys/class/dmi",
+        NULL
+    };
+    int i = 0;
+    while (candidates[i] != NULL) {
+        if(access(candidates[i], F_OK) != -1)
+            return candidates[i];
+        i++;
+    }
+    return NULL;
+}
+
+char *dmi_get_str(const char *id_str) {
+  static struct {
+    char *id;
+    char *path;
+  } tab_dmi_sysfs[] = {
+    /* dmidecode -> sysfs */
+    { "bios-release-date", "id/bios_date" },
+    { "bios-vendor", "id/bios_vendor" },
+    { "bios-version", "id/bios_version" },
+    { "baseboard-product-name", "id/board_name" },
+    { "baseboard-manufacturer", "id/board_vendor" },
+    { "baseboard-version", "id/board_version" },
+    { "baseboard-serial-number", "id/board_serial" },
+    { "baseboard-asset-tag", "id/board_asset_tag" },
+    { "system-product-name", "id/product_name" },
+    { "system-manufacturer", "id/sys_vendor" },
+    { "system-serial-number", "id/product_serial" },
+    { "system-product-family", "id/product_family" },
+    { "system-version", "id/product_version" },
+    { "system-uuid", "product_uuid" },
+    { "chassis-type", "id/chassis_type" },
+    { "chassis-serial-number", "id/chassis_serial" },
+    { "chassis-manufacturer", "id/chassis_vendor" },
+    { "chassis-version", "id/chassis_version" },
+    { "chassis-asset-tag", "id/chassis_asset_tag" },
+    { NULL, NULL }
+  };
+  const gchar *dmi_root = dmi_sysfs_root();
+  gchar *full_path = NULL, *ret = NULL;
+  gboolean spawned;
+  gchar *out, *err;
+
+  int i = 0;
+
+  /* try sysfs first */
+  if (dmi_root) {
+    while (tab_dmi_sysfs[i].id != NULL) {
+      if (strcmp(id_str, tab_dmi_sysfs[i].id) == 0) {
+        full_path = g_strdup_printf("%s/%s", dmi_root, tab_dmi_sysfs[i].path);
+        if (g_file_get_contents(full_path, &ret, NULL, NULL) )
+          goto dmi_str_done;
+        }
+      i++;
+    }
+  }
+
+  /* try dmidecode, but may require root */
+  full_path = g_strconcat("dmidecode -s ", id_str, NULL);
+  spawned = g_spawn_command_line_sync(full_path,
+            &out, &err, &i, NULL);
+  if (spawned) {
+    if (i == 0)
+      ret = out;
+    else
+      g_free(out);
+    g_free(err);
+  }
+
+dmi_str_done:
+  if (ret != NULL) {
+    ret = strend(ret, '\n');
+    ret = g_strstrip(ret);
+    if (strlen(ret) == 0) {
```

Comment:
No need to call `strlen()` if the only thing you want is to see if the string is empty: just use `if (!*ret)` instead.